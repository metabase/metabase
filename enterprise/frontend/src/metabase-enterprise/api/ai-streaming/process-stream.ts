import { match } from "ts-pattern";
import _ from "underscore";

import type { MetabotHistory } from "metabase-types/api";

import {
  type KnownDataPart,
  dataPartSchema,
  finishPartSchema,
  knownDataPartTypes,
  toolCallPartSchema,
  toolResultPartSchema,
} from "./schemas";
import type { JSONValue } from "./types";

const StreamingPartTypeRegistry = {
  TEXT: "0",
  DATA: "2",
  ERROR: "3",
  FINISH_MESSAGE: "d",
  TOOL_CALL: "9",
  TOOL_RESULT: "a",
} as const;

const StreamingPartTypes = Object.values(StreamingPartTypeRegistry);

type StreamingPartType = (typeof StreamingPartTypes)[number];

/**
 * Concatenates all the chunks into a single Uint8Array
 */
function concatChunks(chunks: Uint8Array[]): Uint8Array {
  const totalLength = chunks.reduce((sum, chunk) => sum + chunk.length, 0);
  const concatenatedChunks = new Uint8Array(totalLength);

  let offset = 0;
  for (const chunk of chunks) {
    concatenatedChunks.set(chunk, offset);
    offset += chunk.length;
  }

  return concatenatedChunks;
}

function parseDataStreamPart(line: string) {
  const firstSeparatorIndex = line.indexOf(":");
  if (firstSeparatorIndex === -1) {
    throw new Error("Failed to parse stream string. No separator found.");
  }

  const prefix = line.slice(0, firstSeparatorIndex);
  if (!StreamingPartTypes.includes(prefix as StreamingPartType)) {
    console.warn(`Recieved invalid message code: ${prefix}`);
    return;
  }

  const code = prefix as StreamingPartType;
  const textValue = line.slice(firstSeparatorIndex + 1);
  const jsonValue: JSONValue = JSON.parse(textValue);

  return match(code)
    .with(StreamingPartTypeRegistry.TEXT, (code) => ({
      code,
      name: "text" as const,
      value: jsonValue,
    }))
    .with(StreamingPartTypeRegistry.DATA, (code) => ({
      code,
      name: "data" as const,
      value: dataPartSchema.validateSync(jsonValue, { strict: true }),
    }))
    .with(StreamingPartTypeRegistry.TOOL_CALL, (code) => ({
      code,
      name: "tool_call" as const,
      value: toolCallPartSchema.validateSync(jsonValue, { strict: true }),
    }))
    .with(StreamingPartTypeRegistry.TOOL_RESULT, (code) => ({
      code,
      name: "tool_result" as const,
      value: toolResultPartSchema.validateSync(jsonValue, { strict: true }),
    }))
    .with(StreamingPartTypeRegistry.ERROR, (code) => ({
      code,
      name: "error" as const,
      value: jsonValue,
    }))
    .with(StreamingPartTypeRegistry.FINISH_MESSAGE, (code) => ({
      code,
      name: "finish_message" as const,
      value: finishPartSchema.validateSync(jsonValue, { strict: true }),
    }))
    .exhaustive();
}

type ParsedStreamPart = Exclude<ReturnType<typeof parseDataStreamPart>, void>;
type ParsedStreamPartName = ParsedStreamPart["name"];

function isKnownDataPart(streamPart: ParsedStreamPart): streamPart is Omit<
  Extract<ParsedStreamPart, { name: "data" }>,
  "value"
> & {
  value: KnownDataPart;
} {
  return (
    streamPart.name === "data" &&
    knownDataPartTypes.includes(streamPart.value.type)
  );
}

type AccumulatedStreamParts = {
  toolCalls: (
    | { toolCallId: string; toolName: string; state: "call" }
    | { toolCallId: string; toolName: string; state: "result"; value: unknown }
  )[];
  text: null | string;
  data: unknown[];
  parts: ParsedStreamPart[];
  history: MetabotHistory;
};

function accumulateStreamParts(streamParts: ParsedStreamPart[]) {
  const acc: AccumulatedStreamParts = {
    toolCalls: [],
    data: [],
    text: null,
    parts: streamParts,
    history: [],
  };

  return streamParts.reduce((acc, streamPart, index) => {
    if (streamPart.name === "text") {
      const lastStreamPart = streamParts[index - 1];
      acc.text = `${acc.text ?? ""}${streamPart.value}`;
      if (lastStreamPart?.name === "text") {
        const historyEntry = acc.history.pop();
        acc.history.push({
          ...historyEntry,
          content: historyEntry.content + streamPart.value,
        });
      } else {
        acc.history.push({ role: "assistant", content: streamPart.value });
      }
    }
    if (streamPart.name === "data") {
      acc.data = acc.data.concat(streamPart.value);
    }
    if (streamPart.name === "tool_call") {
      acc.toolCalls.push({ ...streamPart.value, state: "call" });
      acc.history.push({
        role: "assistant",
        "tool-calls": [
          {
            id: streamPart.value.toolCallId,
            name: streamPart.value.toolName,
            arguments: streamPart.value.args,
          },
        ],
      });
    }
    if (streamPart.name === "tool_result") {
      const toolCallId = streamPart.value.toolCallId;
      const index = acc.toolCalls.findIndex((v) => v.toolCallId === toolCallId);

      if (index === -1) {
        throw new Error(
          "Tool Results must be preceded by the tool call with the same toolCallId",
        );
      }

      acc.toolCalls[index] = {
        ...acc.toolCalls[index],
        state: "result",
        value: streamPart.value.result,
      };
      acc.history.push({
        role: "tool",
        content: streamPart.value.result,
        "tool-call-id": streamPart.value.toolCallId,
      });
    }

    return acc;
  }, acc);
}

type StreamPartValue<name extends ParsedStreamPartName> = Extract<
  ParsedStreamPart,
  { name: name }
>["value"];

export type AIStreamingConfig = {
  onTextPart?: (part: StreamPartValue<"text">) => void;
  // callback is only called if this version of the client is aware of the recieved data part type
  onDataPart?: (part: KnownDataPart) => void;
  onToolCallPart?: (part: StreamPartValue<"tool_call">) => void;
  onToolResultPart?: (part: StreamPartValue<"tool_result">) => void;
  onError?: (error: StreamPartValue<"error">) => void;
};

export async function processChatResponse(
  stream: ReadableStream<Uint8Array>,
  config: AIStreamingConfig,
) {
  const parsedStreamParts: ParsedStreamPart[] = [];
  const reader = stream.getReader();
  const decoder = new TextDecoder();
  let chunks: Uint8Array[] = [];

  // eslint-disable-next-line no-constant-condition
  while (true) {
    const { value: chunk } = await reader.read();

    if (chunk) {
      chunks.push(chunk);
      if (chunk[chunk.length - 1] !== "\n".charCodeAt(0)) {
        // if the last character is not a newline, we have not read the whole JSON value
        continue;
      }
    }

    if (chunks.length === 0) {
      break; // we have reached the end of the stream
    }

    const concatenatedChunks = concatChunks(chunks);
    chunks = [];

    const streamParts = _.compact(
      decoder
        .decode(concatenatedChunks, { stream: true })
        .split("\n")
        .filter((line) => line !== "") // splitting leaves an empty string at the end
        .map(parseDataStreamPart),
    );

    for (const streamPart of streamParts) {
      parsedStreamParts.push(streamPart);

      if (streamPart.name === "text") {
        config.onTextPart?.(streamPart.value);
      }
      if (streamPart.name === "data") {
        if (isKnownDataPart(streamPart)) {
          config.onDataPart?.(streamPart.value);
        } else {
          console.warn("Skipping unknown data part:", streamPart);
        }
      }
      if (streamPart.name === "tool_call") {
        config.onToolCallPart?.(streamPart.value);
      }
      if (streamPart.name === "tool_result") {
        config.onToolResultPart?.(streamPart.value);
      }
      if (streamPart.name === "error") {
        config.onError?.(streamPart.value);
      }
    }
  }

  return accumulateStreamParts(parsedStreamParts);
}
