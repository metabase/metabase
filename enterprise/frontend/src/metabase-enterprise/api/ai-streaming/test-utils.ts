import { defer } from "metabase/lib/promise";

async function delay(timeout: number) {
  return new Promise((resolve) => {
    setTimeout(() => resolve(undefined), timeout);
  });
}

export function createPauses<Count extends number>(count: Count) {
  const pauses = new Array(count).fill(null).map(() => defer());
  return pauses as ReturnType<typeof defer>[] & { length: Count };
}

export function createMockReadableStream(
  textChunks: string[] | AsyncGenerator<string, void, unknown>,
  options?: {
    disableAutoInsertNewLines: boolean;
  },
) {
  return new ReadableStream({
    async start(controller) {
      const textEncoder = new TextEncoder();
      try {
        for await (const textChunk of textChunks) {
          const text =
            textChunk + (options?.disableAutoInsertNewLines ? "" : "\n");
          controller.enqueue(textEncoder.encode(text));
        }
      } finally {
        controller.close();
      }
    },
  });
}

export function mockEndpoint<T extends Response>(
  url: string,
  endpointMock: () => Promise<T>,
) {
  const originalFetch = global.fetch;

  // fetch-mock is supposed to work with ReadableStreams, but when passed one
  // the getReader methods ends up as undefined
  return jest
    .spyOn(global, "fetch")
    .mockImplementation((fetchedUrl, ...args) => {
      const isRequestedUrl =
        typeof fetchedUrl === "string" && fetchedUrl.includes(url);

      if (isRequestedUrl) {
        return endpointMock();
      } else {
        return originalFetch(fetchedUrl, ...args);
      }
    });
}

export type MockStreamedEndpointParams =
  | {
      textChunks: string[] | undefined;
      stream?: undefined;
      initialDelay?: number;
    }
  | {
      textChunks?: undefined;
      stream: ReadableStream<any>;
      initialDelay?: number;
    };

export function mockStreamedEndpoint(
  url: string,
  { textChunks, stream, initialDelay = 0 }: MockStreamedEndpointParams,
) {
  return mockEndpoint(url, async () => {
    await delay(initialDelay);
    return {
      status: 202,
      ok: true,
      body:
        stream ||
        (textChunks && createMockReadableStream(textChunks)) ||
        undefined,
    } as any;
  });
}
