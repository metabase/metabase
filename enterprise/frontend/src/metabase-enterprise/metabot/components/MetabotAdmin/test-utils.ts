import type { MetabotId, SuggestedMetabotPrompt } from "metabase-types/api";

export const mockSuggestedPrompts: Record<MetabotId, SuggestedMetabotPrompt[]> =
  {
    "1": [
      {
        id: 1,
        metabot_id: 1,
        prompt: "What is the total revenue for this quarter?",
        model: "metric",
        model_id: 1,
        model_name: "Quarterly Revenue Calculator",
        created_at: "2025-05-15T10:30:00Z",
        updated_at: "2025-05-15T10:30:00Z",
      },
      {
        id: 2,
        metabot_id: 1,
        prompt:
          "Show me the customer acquisition trends over the last 6 months",
        model: "model",
        model_id: 2,
        model_name: "Customer Acquisition Trend Analyzer",
        created_at: "2025-05-15T11:15:00Z",
        updated_at: "2025-05-15T11:15:00Z",
      },
      {
        id: 3,
        metabot_id: 1,
        prompt: "What are our top performing products by sales volume?",
        model: "metric",
        model_id: 3,
        model_name: "Product Performance Ranking",
        created_at: "2025-05-15T14:22:00Z",
        updated_at: "2025-05-16T09:45:00Z",
      },
      {
        id: 4,
        metabot_id: 1,
        prompt: "How has our monthly recurring revenue changed this year?",
        model: "model",
        model_id: 1,
        model_name: "MRR Growth Tracker",
        created_at: "2025-05-16T08:30:00Z",
        updated_at: "2025-05-16T08:30:00Z",
      },
      {
        id: 5,
        metabot_id: 1,
        prompt: "What is our customer churn rate compared to last quarter?",
        model: "metric",
        model_id: 2,
        model_name: "Churn Rate Comparator",
        created_at: "2025-05-16T13:10:00Z",
        updated_at: "2025-05-16T13:10:00Z",
      },
      {
        id: 6,
        metabot_id: 1,
        prompt: "Show me the geographic distribution of our user base",
        model: "model",
        model_id: 4,
        model_name: "Geographic User Distribution Mapper",
        created_at: "2025-05-17T16:45:00Z",
        updated_at: "2025-05-18T10:20:00Z",
      },
      {
        id: 7,
        metabot_id: 1,
        prompt: "What is the average order value for new customers?",
        model: "metric",
        model_id: 4,
        model_name: "New Customer AOV Calculator",
        created_at: "2025-05-18T09:15:00Z",
        updated_at: "2025-05-18T09:15:00Z",
      },
      {
        id: 8,
        metabot_id: 1,
        prompt: "How do our conversion rates vary by traffic source?",
        model: "model",
        model_id: 3,
        model_name: "Traffic Source Conversion Analyzer",
        created_at: "2025-05-18T15:30:00Z",
        updated_at: "2025-05-18T15:30:00Z",
      },
    ],
    "2": [
      {
        id: 1,
        metabot_id: 1,
        prompt: "What is the total revenue for this quarter?",
        model: "metric",
        model_id: 1,
        model_name: "Quarterly Revenue Calculator",
        created_at: "2025-05-15T10:30:00Z",
        updated_at: "2025-05-15T10:30:00Z",
      },
    ],
  };
