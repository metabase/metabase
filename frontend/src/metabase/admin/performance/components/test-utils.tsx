import { setupEnterprisePlugins } from "__support__/enterprise";
import { setupDatabasesEndpoints } from "__support__/server-mocks";
import { setupPerformanceEndpoints } from "__support__/server-mocks/performance";
import { mockSettings } from "__support__/settings";
import { createMockEntitiesState } from "__support__/store";
import { fireEvent, renderWithProviders, screen } from "__support__/ui";
import type { TokenFeatures } from "metabase-types/api";
import { DurationUnit } from "metabase-types/api";
import {
  createMockSettings,
  createMockTokenFeatures,
} from "metabase-types/api/mocks";
import {
  createMockCacheConfig,
  createMockCacheConfigWithDoNotCacheStrategy,
  createMockCacheConfigWithDurationStrategy,
  createMockCacheConfigWithTTLStrategy,
} from "metabase-types/api/mocks/performance";
import { createSampleDatabase } from "metabase-types/api/mocks/presets";
import { createMockState } from "metabase-types/store/mocks";

import { StrategyEditorForDatabases } from "./StrategyEditorForDatabases";

export interface SetupOpts {
  hasEnterprisePlugins?: boolean;
  tokenFeatures?: Partial<TokenFeatures>;
}

export const setup = ({
  hasEnterprisePlugins,
  tokenFeatures = {},
}: SetupOpts = {}) => {
  const storeInitialState = createMockState({
    entities: createMockEntitiesState({}),
    settings: mockSettings(
      createMockSettings({
        "token-features": createMockTokenFeatures(tokenFeatures),
      }),
    ),
  });

  if (hasEnterprisePlugins) {
    setupEnterprisePlugins();
  }

  const cacheConfigs = [
    createMockCacheConfigWithTTLStrategy({ model_id: 1 }),
    createMockCacheConfigWithDoNotCacheStrategy({ model_id: 2 }),
    createMockCacheConfigWithDurationStrategy({ model_id: 3 }),
    createMockCacheConfig({
      model: "root",
      model_id: 0,
      strategy: { type: "duration", duration: 1, unit: DurationUnit.Hours },
    }),
  ];
  setupPerformanceEndpoints(cacheConfigs);

  const databases = Array.from({ length: 4 }, (_, i) =>
    createSampleDatabase({ id: i + 1, name: `Database ${i + 1}`, tables: [] }),
  );
  setupDatabasesEndpoints(databases);

  return renderWithProviders(<StrategyEditorForDatabases />, {
    storeInitialState,
  });
};

export const getSaveButton = async () =>
  await screen.findByTestId("strategy-form-submit-button");

export const changeInput = async (
  label: RegExp,
  expectedPlaceholder: number,
  value: number,
) => {
  const input = (await screen.findByLabelText(label)) as HTMLInputElement;
  expect(input).toHaveAttribute("placeholder", expectedPlaceholder.toString());
  fireEvent.change(input, { target: { value } });
  expect(input).toHaveValue(value);
};
