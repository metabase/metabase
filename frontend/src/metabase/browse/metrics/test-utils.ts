import {
  createMockRecentCollectionItem,
  createMockSearchResult,
} from "metabase-types/api/mocks";

import type { MetricResult, RecentMetric } from "./types";

export const createMockMetricResult = (
  metric: Partial<MetricResult> = {},
): MetricResult =>
  createMockSearchResult({ ...metric, model: "metric" }) as MetricResult;

export const createMockRecentMetric = (
  metric: Partial<RecentMetric>,
): RecentMetric =>
  createMockRecentCollectionItem({
    ...metric,
    model: "metric",
  }) as RecentMetric;
