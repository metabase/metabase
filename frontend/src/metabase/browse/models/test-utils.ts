import type { RecentCollectionItem } from "metabase-types/api";
import {
  createMockRecentCollectionItem,
  createMockSearchResult,
} from "metabase-types/api/mocks";

import type { ModelResult, RecentModel } from "./types";

export const createMockModelResult = (
  model: Partial<ModelResult> = {},
): ModelResult =>
  createMockSearchResult({ ...model, model: "dataset" }) as ModelResult;

export const createMockRecentModel = (
  model: Partial<RecentCollectionItem>,
): RecentModel =>
  createMockRecentCollectionItem({ ...model, model: "dataset" }) as RecentModel;
