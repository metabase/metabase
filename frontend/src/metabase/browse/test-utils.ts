import type { ModelResult } from "metabase-types/api";
import { createMockSearchResult } from "metabase-types/api/mocks";

export const createMockModelResult = (
  model: Partial<ModelResult>,
): ModelResult =>
  createMockSearchResult({ ...model, model: "dataset" }) as ModelResult;
