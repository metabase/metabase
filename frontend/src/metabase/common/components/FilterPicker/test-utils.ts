/* istanbul ignore file */
import dayjs from "dayjs";
import { createMockEntitiesState } from "__support__/store";
import { checkNotNull } from "metabase/lib/types";
import { getMetadata } from "metabase/selectors/metadata";
import type { FieldValuesType } from "metabase-types/api";
import { createMockField, createMockSegment } from "metabase-types/api/mocks";
import {
  createSampleDatabase,
  createOrdersTable,
  createPeopleTable,
  createProductsTable,
  ORDERS,
  ORDERS_ID,
  PEOPLE_ID,
  PRODUCTS_ID,
} from "metabase-types/api/mocks/presets";
import { createMockState } from "metabase-types/store/mocks";
import * as Lib from "metabase-lib";
import { TYPE } from "metabase-lib/types/constants";
import {
  createQuery as _createQuery,
  columnFinder,
} from "metabase-lib/test-helpers";

const SEGMENT_1 = createMockSegment({
  id: 1,
  table_id: ORDERS_ID,
  name: "Discounted",
  description: "Discounted",
  definition: {
    "source-table": ORDERS_ID,
    filter: ["not-null", ["field", ORDERS.DISCOUNT, null]],
  },
});

const SEGMENT_2 = createMockSegment({
  id: 2,
  table_id: ORDERS_ID,
  name: "Many items",
  description: "Orders with more than 5 items",
  definition: {
    "source-table": ORDERS_ID,
    filter: [">", ["field", ORDERS.QUANTITY, null], 20],
  },
});

const BOOLEAN_FIELD = createMockField({
  id: 100,
  table_id: PEOPLE_ID,
  name: "IS_ACTIVE",
  display_name: "Is Active",

  base_type: TYPE.Boolean,
  effective_type: TYPE.Boolean,
  semantic_type: null,
});

const STRING_FIELD_NO_VALUES = createMockField({
  id: 101,
  table_id: PRODUCTS_ID,
  name: "DESCRIPTION",
  display_name: "Description",

  base_type: TYPE.Text,
  effective_type: TYPE.Text,
  semantic_type: null,

  has_field_values: "none",
});

const TIME_FIELD = createMockField({
  id: 102,
  table_id: ORDERS_ID,
  name: "TIME",
  display_name: "Time",

  base_type: TYPE.Time,
  effective_type: TYPE.Time,
  semantic_type: null,
});

const _ordersFields = createOrdersTable().fields?.filter(checkNotNull) ?? [];
const _peopleFields = createPeopleTable().fields?.filter(checkNotNull) ?? [];
const _productsFields =
  createProductsTable().fields?.filter(checkNotNull) ?? [];

const database = createSampleDatabase({
  tables: [
    createOrdersTable({
      fields: [..._ordersFields, TIME_FIELD],
      segments: [SEGMENT_1, SEGMENT_2],
    }),
    createPeopleTable({ fields: [..._peopleFields, BOOLEAN_FIELD] }),
    createProductsTable({
      fields: [..._productsFields, STRING_FIELD_NO_VALUES],
    }),
  ],
});

export const storeInitialState = createMockState({
  entities: createMockEntitiesState({
    databases: [database],
    segments: [SEGMENT_1, SEGMENT_2],
  }),
});

export const metadata = getMetadata(storeInitialState);

export function createQuery() {
  return _createQuery({ metadata });
}

export function createFilteredQuery(
  initialQuery: Lib.Query,
  clause: Lib.ExpressionClause | Lib.SegmentMetadata,
) {
  const query = Lib.filter(initialQuery, 0, clause);
  const [filter] = Lib.filters(query, 0);
  const column = Lib.filterParts(query, 0, filter)?.column;
  return { query, filter, column };
}

function findFilteredColumn(
  query: Lib.Query,
  tableName: string,
  columnName: string,
) {
  const columns = Lib.filterableColumns(query, 0);
  const findColumn = columnFinder(query, columns);
  return findColumn(tableName, columnName);
}

export function findBooleanColumn(query: Lib.Query) {
  return findFilteredColumn(query, "PEOPLE", "IS_ACTIVE");
}

type BooleanFilterQueryOpts = Partial<Lib.BooleanFilterParts> & {
  query?: Lib.Query;
  column?: Lib.ColumnMetadata;
};

export function createQueryWithBooleanFilter({
  query = createQuery(),
  column = findBooleanColumn(query),
  operator = "=",
  values = [true],
}: BooleanFilterQueryOpts = {}) {
  const clause = Lib.booleanFilterClause({ operator, column, values });
  return createFilteredQuery(query, clause);
}

export function findLatitudeColumn(query: Lib.Query) {
  return findFilteredColumn(query, "PEOPLE", "LATITUDE");
}

export function findLongitudeColumn(query: Lib.Query) {
  return findFilteredColumn(query, "PEOPLE", "LONGITUDE");
}

type CoordinateFilterQueryOpts = Partial<Lib.CoordinateFilterParts> & {
  query?: Lib.Query;
  column?: Lib.ColumnMetadata;
};

export function createQueryWithCoordinateFilter({
  query = createQuery(),
  column = findLatitudeColumn(query),
  operator = "=",
  values = [0],
  ...parts
}: CoordinateFilterQueryOpts = {}) {
  const clause = Lib.coordinateFilterClause({
    operator,
    column,
    values,
    ...parts,
  });
  return createFilteredQuery(query, clause);
}

export function findNumericColumn(query: Lib.Query) {
  return findFilteredColumn(query, "ORDERS", "TOTAL");
}

type NumberFilterQueryOpts = Partial<Lib.NumberFilterParts> & {
  query?: Lib.Query;
  column?: Lib.ColumnMetadata;
};

export function createQueryWithNumberFilter({
  query = createQuery(),
  column = findNumericColumn(query),
  operator = "=",
  values = [0],
}: NumberFilterQueryOpts = {}) {
  const clause = Lib.numberFilterClause({ operator, column, values });
  return createFilteredQuery(query, clause);
}

export function findStringColumn(
  query: Lib.Query,
  { fieldValues = "none" }: { fieldValues?: FieldValuesType } = {},
) {
  const fieldNameMap = {
    none: "DESCRIPTION",
    list: "CATEGORY",
    search: "VENDOR",
  };
  return findFilteredColumn(query, "PRODUCTS", fieldNameMap[fieldValues]);
}

type StringFilterOpts = Partial<Lib.StringFilterParts> & {
  query?: Lib.Query;
  column?: Lib.ColumnMetadata;
};

export function createQueryWithStringFilter({
  query = createQuery(),
  column = findStringColumn(query),
  operator = "=",
  values = [""],
  options = {},
}: StringFilterOpts = {}) {
  const clause = Lib.stringFilterClause({ operator, column, values, options });
  return createFilteredQuery(query, clause);
}

export function findDateColumn(query: Lib.Query) {
  return findFilteredColumn(query, "PEOPLE", "BIRTH_DATE");
}

export function findDateTimeColumn(query: Lib.Query) {
  return findFilteredColumn(query, "ORDERS", "CREATED_AT");
}

type SpecificDateFilterOpts = Partial<Lib.SpecificDateFilterParts> & {
  query?: Lib.Query;
  column?: Lib.ColumnMetadata;
};

export function createQueryWithSpecificDateFilter({
  query = createQuery(),
  column = findDateTimeColumn(query),
  operator = "=",
  values = [new Date(2020, 1, 15)],
}: SpecificDateFilterOpts = {}) {
  const clause = Lib.specificDateFilterClause(query, 0, {
    operator,
    column,
    values,
  });
  return createFilteredQuery(query, clause);
}

type RelativeDateFilterOpts = Partial<Lib.RelativeDateFilterParts> & {
  query?: Lib.Query;
  column?: Lib.ColumnMetadata;
};

export function createQueryWithRelativeDateFilter({
  query = createQuery(),
  column = findDateTimeColumn(query),
  value = -20,
  bucket = "day",
  offsetValue = null,
  offsetBucket = null,
  options = {},
}: RelativeDateFilterOpts = {}) {
  const clause = Lib.relativeDateFilterClause({
    column,
    value,
    bucket,
    offsetValue,
    offsetBucket,
    options,
  });
  return createFilteredQuery(query, clause);
}

type ExcludeDateFilterOpts = Partial<Lib.ExcludeDateFilterParts> & {
  query?: Lib.Query;
  column?: Lib.ColumnMetadata;
};

export function createQueryWithExcludeDateFilter({
  query = createQuery(),
  column = findDateTimeColumn(query),
  operator = "!=",
  values = [1],
  bucket = "day-of-week",
}: ExcludeDateFilterOpts = {}) {
  const clause = Lib.excludeDateFilterClause(query, 0, {
    column,
    operator,
    values,
    bucket,
  });
  return createFilteredQuery(query, clause);
}

export function findTimeColumn(query: Lib.Query) {
  return findFilteredColumn(query, "ORDERS", "TIME");
}

type TimeFilterOpts = Partial<Lib.TimeFilterParts> & {
  query?: Lib.Query;
  column?: Lib.ColumnMetadata;
};

export function createQueryWithTimeFilter({
  query = createQuery(),
  column = findTimeColumn(query),
  operator = ">",
  values = [dayjs().startOf("day").toDate()],
}: TimeFilterOpts = {}) {
  const clause = Lib.timeFilterClause({ operator, column, values });
  return createFilteredQuery(query, clause);
}
