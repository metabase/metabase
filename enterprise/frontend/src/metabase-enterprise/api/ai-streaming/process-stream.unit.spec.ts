import { processChatResponse } from "./process-stream";
import { createMockReadableStream } from "./test-utils";

const getMockedCallbacks = () => ({
  onTextPart: jest.fn(),
  onDataPart: jest.fn(),
  onToolCallPart: jest.fn(),
  onToolResultPart: jest.fn(),
  onError: jest.fn(),
});

const expectNoStreamedError = {
  onError: (error: unknown) => {
    expect(error).toBeUndefined();
  },
};

const mockSuccessStreamData = [
  `0:"You, but don't tell anyone."`,
  `2:{"type":"state","version":1,"value":{"queries":{}}}`,
  `9:{"toolCallId":"x","toolName":"x","args":""}`,
  `a:{"toolCallId":"x","result":""}`,
  `d:{"finishReason":"stop","usage":{"promptTokens":4916,"completionTokens":8}}`,
];
const getMockSuccessStream = () =>
  createMockReadableStream(mockSuccessStreamData);

const getMockErrorStream = () => createMockReadableStream([`3:{}`]);

describe("processChatResponse", () => {
  it("should be able to process a valid stream", async () => {
    const mockStream = getMockSuccessStream();
    const result = await processChatResponse(mockStream, expectNoStreamedError);
    expect(result).toMatchSnapshot();
  });

  it("should call callbacks for relevant chunk types", async () => {
    const mockSuccessStream = getMockSuccessStream();
    const config = getMockedCallbacks();

    await processChatResponse(mockSuccessStream, config);
    expect(config.onTextPart).toHaveBeenCalled();
    expect(config.onDataPart).toHaveBeenCalled();
    expect(config.onToolCallPart).toHaveBeenCalled();
    expect(config.onToolResultPart).toHaveBeenCalled();
    expect(config.onError).not.toHaveBeenCalled();

    const mockErrorStream = getMockErrorStream();
    try {
      await processChatResponse(mockErrorStream, config);
    } catch (_) {}
    expect(config.onError).toHaveBeenCalled();
  });

  it("should ignore unknown data parts", async () => {
    const mockStream = createMockReadableStream([
      `2:{"type":"__some_futurist_data__","version":1,"value":"hi"}`,
      `d:{"finishReason":"stop","usage":{"promptTokens":4916,"completionTokens":8}}`,
    ]);
    const config = getMockedCallbacks();

    const result = await processChatResponse(mockStream, config);
    // callbacks shouldn't have been triggered
    expect(config.onError).not.toHaveBeenCalled();
    expect(config.onDataPart).not.toHaveBeenCalled();
    // we should keep track of the info in the result
    expect(result.data).toEqual([
      { type: "__some_futurist_data__", value: "hi", version: 1 },
    ]);
  });

  it("should ignore unknown part types", async () => {
    const mockStream = createMockReadableStream([`x:"UNKNOWN PART TYPE"`]);
    const config = getMockedCallbacks();
    await expect(processChatResponse(mockStream, config)).resolves.toBeTruthy();
    expect(config.onError).not.toHaveBeenCalled();
  });

  it("should error if there is no part type", async () => {
    const mockStream = createMockReadableStream([`data-without-part-type`]);
    await expect(
      processChatResponse(mockStream, expectNoStreamedError),
    ).rejects.toBeTruthy();
  });

  it("should error if a tool result is returned without a preceeding tool call", async () => {
    const mockStream = createMockReadableStream([
      `a:{"toolCallId":"x","result":""}`,
      `d:{"finishReason":"stop","usage":{"promptTokens":4916,"completionTokens":8}}`,
    ]);
    await expect(
      processChatResponse(mockStream, expectNoStreamedError),
    ).rejects.toBeTruthy();
  });

  it("should handle messages across multiple chunks", async () => {
    const mockStream = createMockReadableStream(
      [
        `0:"You, but `,
        `don't tell anyone."\n`,
        `d:{"finishReason":"stop","usage":{"promptTokens":4916,"completionTokens":8}}\n`,
      ],
      { disableAutoInsertNewLines: true },
    );

    const result = await processChatResponse(mockStream, expectNoStreamedError);
    expect(result.text).toEqual("You, but don't tell anyone.");
    expect(result.history).toEqual([
      { content: "You, but don't tell anyone.", role: "assistant" },
    ]);
  });
});
