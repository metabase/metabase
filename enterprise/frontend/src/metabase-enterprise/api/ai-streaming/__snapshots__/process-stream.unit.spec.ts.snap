// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processChatResponse should be able to process a valid stream 1`] = `
{
  "data": [
    {
      "type": "state",
      "value": {
        "queries": {},
      },
      "version": 1,
    },
  ],
  "history": [
    {
      "content": "You, but don't tell anyone.",
      "role": "assistant",
    },
    {
      "role": "assistant",
      "tool-calls": [
        {
          "arguments": "",
          "id": "x",
          "name": "x",
        },
      ],
    },
    {
      "content": "",
      "role": "tool",
      "tool-call-id": "x",
    },
  ],
  "parts": [
    {
      "code": "0",
      "name": "text",
      "value": "You, but don't tell anyone.",
    },
    {
      "code": "2",
      "name": "data",
      "value": {
        "type": "state",
        "value": {
          "queries": {},
        },
        "version": 1,
      },
    },
    {
      "code": "9",
      "name": "tool_call",
      "value": {
        "args": "",
        "toolCallId": "x",
        "toolName": "x",
      },
    },
    {
      "code": "a",
      "name": "tool_result",
      "value": {
        "result": "",
        "toolCallId": "x",
      },
    },
    {
      "code": "d",
      "name": "finish_message",
      "value": {
        "finishReason": "stop",
        "usage": {
          "completionTokens": 8,
          "promptTokens": 4916,
        },
      },
    },
  ],
  "text": "You, but don't tell anyone.",
  "toolCalls": [
    {
      "args": "",
      "state": "result",
      "toolCallId": "x",
      "toolName": "x",
      "value": "",
    },
  ],
}
`;
